File: cometchat_calls_sdk/ios/Classes/Wrapper/CometChatCallsPlatformViewFactory.swift
import Foundation
import Flutter
import CometChatCallsSDK
import UIKit

class CometChatCallsPlatformViewFactory: NSObject, FlutterPlatformViewFactory, CallsEventsDelegate {
    private var messenger: FlutterBinaryMessenger?
    private var callToken: String?
    private var call: FlutterMethodCall?
    private var result: FlutterResult?
    private var view = UIView()
    private var eventSink: FlutterEventSink?
    
    //This fun will be called by FlutterPlatformViewFactory for creating native view
    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        print("iOS CometChatCallsPlatformViewFactory create: here")
        return CometChatCallsPlatformView(
            frame: frame,
            viewIdentifier: viewId,
            arguments: args,
            view: view)
    }
    
    //This will initialise the data for the call
    func initData(messenger: FlutterBinaryMessenger?,
                callToken: String,
                call: FlutterMethodCall,
                result: @escaping (FlutterResult),
                eventSink: FlutterEventSink?
    ){
        self.messenger = messenger
        self.callToken = callToken
        self.call = call
        self.result = result
        self.eventSink = eventSink
        
        print("iOS CometChatCallsPlatformViewFactory setData: here")
    }
    
    //this func will set up call settings form arg and start the call
    func startNormalCall(){
        
        DispatchQueue.main.async {
            
            //assigning new UIView for second time call start 
            self.view = UIView()
            
            let arg = self.call?.arguments as? [String: Any]
            
            //View Settings
            let videoSettings = MainVideoContainerSetting()
            videoSettings.setMainVideoAspectRatio(arg?["videoFit"] as? String ?? "contain")
            videoSettings.setFullScreenButtonParams(
                ((arg?["fullScreenButtonVisibility"] as? Bool) ?? true),
                (((arg?["fullScreenButtonPosition"]) as? String) ?? "bottom-right")
            )
            
            videoSettings.setZoomButtonParams(
                ((arg?["zoomButtonVisibility"] as? Bool) ?? true),
                (((arg?["zoomButtonPosition"]) as? String) ?? "bottom-right")
                
            )
            
            videoSettings.setNameLabelParams(
                ((arg?["userListButtonVisibility"] as? Bool) ?? true),
                (((arg?["userListButtonPosition"]) as? String) ?? "bottom-right"),
                (((arg?["nameLabelColor"]) as? String) ?? "#333333")
            )
            
            videoSettings.setUserListButtonParams(
                ((arg?["userListButtonPosition"] as? Bool) ?? true),
                (((arg?["userListButtonVisibility"]) as? String) ?? "bottom-right")
                
            )
            
            print("iOS videoSettings: \(videoSettings.getDictionaryForVideoSetting())")
            
            //Main Call Settings
            let callSettings = CometChatCalls.callSettingsBuilder
                .setDefaultLayout((arg?["defaultLayout"] as? Bool) ?? true)
                .setIsAudioOnly((arg?["isAudioOnly"] as? Bool) ?? false)
                .setIsSingleMode((arg?["isSingleMode"] as? Bool) ?? false)
                .setShowSwitchToVideoCall((arg?["showSwitchToVideoCall"] as? Bool) ?? false)
                .setEnableVideoTileClick((arg?["enableVideoTileClick"] as? Bool) ?? true)
                .setEnableDraggableVideoTile((arg?["enableDraggableVideoTile"] as? Bool) ?? true)
                .setEndCallButtonDisable((arg?["endCallButtonDisable"] as? Bool) ?? false)
                .setShowRecordingButton((arg?["showRecordingButton"] as? Bool) ?? true)
                .setSwitchCameraButtonDisable((arg?["switchCameraButtonDisable"] as? Bool) ?? false)
                .setStartRecordingOnCallStart((arg?["startRecordingOnCallStart"] as? Bool) ?? false)
                .setMuteAudioButtonDisable((arg?["muteAudioButtonDisable"] as? Bool) ?? false)
                .setPauseVideoButtonDisable((arg?["pauseVideoButtonDisable"] as? Bool) ?? false)
                .setAudioModeButtonDisable((arg?["audioModeButtonDisable"] as? Bool) ?? false)
                .setStartAudioMuted((arg?["startAudioMuted"] as? Bool) ?? true)
                .setStartVideoMuted((arg?["startVideoMuted"] as? Bool) ?? false)
                .setAvatarMode((arg?["avatarMode"] as? NSString) ?? "CIRCLE")
                .setMode((arg?["mode"] as? NSString) ?? "DEFAULT")
                .setDefaultAudioMode((arg?["defaultAudioMode"] as? NSString) ?? "SPEAKER")
                .setVideoSettings(videoSettings.getDictionaryForVideoSetting())
                .setDelegate(self)
                .build()
            
            print("CometChatiOS: Starting Section...")
            guard let callToken = self.callToken else {
                print("CometChatiOS: Error no Auth token")
                return
            }
            
            print("CometChatiOS call token: \(callToken)")
            
            CometChatCalls.startSession(callToken: callToken, callSetting: callSettings, view: self.view) { response in
                print("CometChatiOS: Session response \(response)")
                self.result?(GetMappedObject.resultMap(msg: response, methodName: "success"))
            } onError: { error in
                print("CometChatiOS: Session start failed with error \(String(describing: error?.description))")
                self.result?(FlutterError(code: error?.errorCode ?? "START SESSION ERROR", message: error?.description, details: error?.description))
            }
        }
    }
    
    func startPresenterModeCall() {
        
        DispatchQueue.main.async {
            
            //assigning new UIView for second time call start
            self.view = UIView()
            
            let arg = self.call?.arguments as? [String: Any]
            
            //Presentation Call Settings
            let presentationSettings = CometChatCalls.presentationSettingsBuilder
                .setDefaultLayout((arg?["defaultLayout"] as? Bool) ?? true)
                .setIsAudioOnly((arg?["isAudioOnly"] as? Bool) ?? false)
                .setEndCallButtonDisable((arg?["endCallButtonDisable"] as? Bool) ?? false)
                .setShowRecordingButton((arg?["showRecordingButton"] as? Bool) ?? true)
                .setSwitchCameraButtonDisable((arg?["switchCameraButtonDisable"] as? Bool) ?? false)
                .setMuteAudioButtonDisable((arg?["muteAudioButtonDisable"] as? Bool) ?? false)
                .setPauseVideoButtonDisable((arg?["pauseVideoButtonDisable"] as? Bool) ?? false)
                .setAudioModeButtonDisable((arg?["audioModeButtonDisable"] as? Bool) ?? false)
                .setStartAudioMuted((arg?["startAudioMuted"] as? Bool) ?? true)
                .setStartVideoMuted((arg?["startVideoMuted"] as? Bool) ?? false)
                .setDefaultAudioMode((arg?["defaultAudioMode"] as? NSString) ?? "BLUETOOTH")
                .setIsPresenter(arg?["isPresenter"] as? Bool ?? false)
                .setDelegate(self)
                .build()
            
            print("CometChatiOS: Starting Section...")
            guard let callToken = self.callToken else {
                print("CometChatiOS: Error no Auth token")
                return
            }
            
            print("CometChatiOS call token: \(callToken)")
            
            CometChatCalls.joinPresentation(callToken: callToken, presenterSettings: presentationSettings, view: self.view) { response in
                print("CometChatiOS: Session response \(response)")
                self.result?(GetMappedObject.resultMap(msg: response, methodName: "success"))
            } onError: { error in
                print("CometChatiOS: Session start failed with error \(String(describing: error?.errorDescription))")
                self.result?(FlutterError(code: error?.errorCode ?? "START PRESENTATION ERROR", message: error?.description, details: error?.description))
            }
        }
        
    }
    
    //MARK: CALL EVENT DELEGATE
    func onCallEnded() {
        print("CometChatCalls-Ios: call ended")
        eventSink?(GetMappedObject.boolMap(flag: true, methodName: "onCallEnded"))
    }
    
    func onCallEndButtonPressed() {
        print("CometChatCalls-Ios: onCallEndButtonPressed")
        eventSink?(GetMappedObject.boolMap(flag: true, methodName: "onCallEndButtonPressed"))
    }
    
    func onUserJoined(user: NSDictionary) {
        print("CometChatCalls-Ios: onUserJoined")
        eventSink?(GetMappedObject.rtcUserMap(rtcUser: user, methodName: "onUserJoined"))
    }
    
    func onUserLeft(user: NSDictionary) {
        print("CometChatCalls-Ios: onUserLeft")
        eventSink?(GetMappedObject.rtcUserMap(rtcUser: user, methodName: "onUserLeft"))
    }
    
    func onUserListChanged(userList: NSArray) {
        print("CometChatCalls-Ios: onUserListChanged")
        eventSink?(GetMappedObject.rtcUserListMap(rtcUserList: userList))
    }
    
    func onAudioModeChanged(audioModeList: NSArray) {
        print("CometChatCalls-Ios: onAudioModeChanged")
        eventSink?(GetMappedObject.audioModeListMap(audioModeList: audioModeList, methodName: "onAudioModeChanged") ?? nil)
        
    }
    
    func onCallSwitchedToVideo(info: NSDictionary) {
        print("CometChatCalls-Ios: onCallSwitchedToVideo")
        eventSink?(GetMappedObject.callSwitchRequestInfoMap(callSwitchRequestInfo: info, methodName: "onCallSwitchedToVideo"))
    }
    
    func onUserMuted(info: NSDictionary) {
        print("CometChatCalls-Ios: onUserMuted \(info)")
        eventSink?(GetMappedObject.rtcMutedUserMap(rtcMutedUser: info, methodName: "onUserMuted   "))
    }
    
    func onRecordingToggled(info: NSDictionary) {
        print("CometChatCalls-Ios: onRecordingToggled \(info)")
        eventSink?(GetMappedObject.rtcRecordingInfoMap(rtcRecordingInfo: info, methodName: "onRecordingToggled"))
    }

}

File: cometchat_calls_sdk/android/src/main/kotlin/com/cometchatcalls/cometchatcalls_plugin/wrapper/CometChatCallsPlatformViewFactory.kt
package com.cometchatcalls.cometchatcalls_plugin.wrapper

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.widget.RelativeLayout
import com.cometchat.calls.core.CometChatCalls
import com.cometchat.calls.exceptions.CometChatException
import com.cometchat.calls.listeners.CometChatCallsEventsListener
import com.cometchat.calls.model.AudioMode
import com.cometchat.calls.model.CallSwitchRequestInfo
import com.cometchat.calls.model.MainVideoContainerSetting
import com.cometchat.calls.model.RTCMutedUser
import com.cometchat.calls.model.RTCRecordingInfo
import com.cometchat.calls.model.RTCUser
import com.cometchatcalls.cometchatcalls_plugin.helper.GetMappedObject
import com.cometchatcalls.cometchatcalls_plugin.helper.PluginConstants
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory


/**
 * Created by Rohit Giri on 21/03/23.
 */

class CometChatCallsPlatformViewFactory : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    val TAG = javaClass.simpleName

    companion object{
        lateinit var activity: Activity
        lateinit var call: MethodCall
        lateinit var result: MethodChannel.Result
        lateinit var callsEventsStreamSink: EventChannel.EventSink
        private lateinit var videoContainer: RelativeLayout
        private lateinit var parentContainer: RelativeLayout
        private val uiThreadHandler: Handler = Handler(Looper.getMainLooper())
    }

    fun initVariables(
        activity: Activity,
        call: MethodCall,
        result: MethodChannel.Result,
        callsEventsStreamSink: EventChannel.EventSink
    ){
        CometChatCallsPlatformViewFactory.activity = activity
        CometChatCallsPlatformViewFactory.call = call
        CometChatCallsPlatformViewFactory.result = result
        CometChatCallsPlatformViewFactory.callsEventsStreamSink = callsEventsStreamSink
    }

    fun initCallView() {
        /*parentContainer = RelativeLayout(activity)
        val inflater = LayoutInflater.from(activity)
        val layout = inflater.inflate(R.layout.video_container, null, false) as RelativeLayout
        videoContainer = layout.findViewById(R.id.videoContainer) as RelativeLayout
        parentContainer.addView(layout)*/

        videoContainer = RelativeLayout(activity)

        /*val linearLayout = LinearLayout(activity)
        linearLayout.orientation = LinearLayout.HORIZONTAL
        linearLayout.gravity = Gravity.CENTER
        linearLayout.layoutParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        val layoutParams2 : RelativeLayout.LayoutParams  = linearLayout.layoutParams as  (RelativeLayout.LayoutParams)
        layoutParams2.addRule(RelativeLayout.CENTER_IN_PARENT)

        val progressBar = ProgressBar(activity)
        progressBar.layoutParams = RelativeLayout.LayoutParams(
            55,
            55
        )
        linearLayout.addView(progressBar)

        val textView = TextView(activity)
        textView.text = "Connecting..."
        textView.setTextColor(Color.parseColor("#FFFFFF"))
        val params: RelativeLayout.LayoutParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        params.setMargins(25, 0, 0, 0)
        textView.layoutParams = params
        linearLayout.addView(textView)

        val defaultBackgroundColor: Int = Color.parseColor("#333333")
        videoContainer.setBackgroundColor(defaultBackgroundColor)
        videoContainer.addView(linearLayout)*/

        val videoSettings = MainVideoContainerSetting()
        videoSettings.setVideoStreamProp(call.argument("videoFit") ?: "contain")
        videoSettings.setfullScreenButtonProps(
            call.argument("fullScreenButtonPosition") ?: "bottom-right",
            call.argument<Boolean>("fullScreenButtonVisibility") ?: true
        )
        videoSettings.setUserListButtonProps(
            call.argument("userListButtonPosition") ?: "bottom-right",
            call.argument<Boolean>("userListButtonVisibility") ?: true
        )
        videoSettings.setZoomButtonProps(
            call.argument("zoomButtonPosition") ?: "bottom-right",
            call.argument<Boolean>("zoomButtonVisibility") ?: true
        )
        videoSettings.setNameLabelProps(
            call.argument("nameLabelPosition") ?: "bottom-left",
            call.argument<Boolean>("nameLabelVisibility") ?: true,
            call.argument("nameLabelColor") ?: "#333333"
        )
        val settings = CometChatCalls.CallSettingsBuilder(activity)
            .setDefaultLayoutEnable(call.argument<Boolean>("defaultLayout") ?: true)
            .setIsAudioOnly(call.argument<Boolean>("isAudioOnly") ?: false)
            .setSingleMode(call.argument<Boolean>("isSingleMode") ?: false)
            .showSwitchToVideoCallButton(call.argument<Boolean>("showSwitchToVideoCall") ?: false)
            .enableVideoTileClick(call.argument<Boolean>("enableVideoTileClick") ?: true)
            .enableVideoTileDrag(call.argument<Boolean>("enableDraggableVideoTile") ?: true)
            .showEndCallButton(call.argument<Boolean>("endCallButtonDisable") ?: false)
            .showRecordingButton(call.argument<Boolean>("showRecordingButton") ?: false)
            .autoRecordOnCallStart(call.argument<Boolean>("startRecordingOnCallStart") ?: false)
            .showSwitchCameraButton(call.argument<Boolean>("switchCameraButtonDisable") ?: false)
            .showMuteAudioButton(call.argument<Boolean>("muteAudioButtonDisable") ?: false)
            .showPauseVideoButton(call.argument<Boolean>("pauseVideoButtonDisable") ?: false)
            .showAudioModeButton(call.argument<Boolean>("audioModeButtonDisable") ?: false)
            .startWithAudioMuted(call.argument<Boolean>("startAudioMuted") ?: false)
            .startWithVideoMuted(call.argument<Boolean>("startVideoMuted") ?: false)
            .setAvatarMode(call.argument<String>("avatarMode") ?: "CIRCLE")
            .setMode(call.argument<String>("mode") ?: "DIRECT")
            .setDefaultAudioMode(call.argument<String>("defaultAudioMode") ?: "SPEAKER")
            .setMainVideoContainerSetting(videoSettings)
            .setEventListener(object : CometChatCallsEventsListener {
                override fun onCallEnded() {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.booleanMap(true, "onCallEnded"))
                    }
                }

                override fun onCallEndButtonPressed() {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.booleanMap(true, "onCallEndButtonPressed"))
                    }
                }

                override fun onUserJoined(p0: RTCUser?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcUserMap(p0, "onUserJoined"))
                    }
                }

                override fun onUserLeft(p0: RTCUser?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcUserMap(p0, "onUserLeft"))
                    }
                }

                override fun onUserListChanged(p0: ArrayList<RTCUser>?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcUserListMap(p0, "onUserListChanged"))
                    }
                }

                override fun onAudioModeChanged(p0: ArrayList<AudioMode>?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.audioModeListMap(p0, "onAudioModeChanged"))
                    }
                }

                override fun onCallSwitchedToVideo(p0: CallSwitchRequestInfo?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.callSwitchRequestInfoMap(p0, "onCallSwitchedToVideo"))
                    }
                }

                override fun onUserMuted(p0: RTCMutedUser?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcMutedUserMap(p0, "onUserMuted"))
                    }
                }

                override fun onRecordingToggled(p0: RTCRecordingInfo?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcRecordingInfoMap(p0, "onRecordingToggled"))
                    }
                }

                override fun onError(p0: CometChatException?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.cometchatExceptionMap(p0, "onError"))
                    }
                }
            })
            .build()

        val callToken: String ?= call.argument<String>(PluginConstants.CometchatCallsStartSessionArgumentsKeys.CALL_TOKEN)

        CometChatCalls.startSession(callToken, settings, videoContainer, object : CometChatCalls.CallbackListener<String>(){
            override fun onSuccess(p0: String) {
                uiThreadHandler.post {
                    result.success(GetMappedObject.stringToMap(p0, PluginConstants.Method.SUCCESS))
                }
            }

            override fun onError(e: CometChatException) {
                uiThreadHandler.post {
                    result.error(e.code, e.message, e.details)
                }
            }
        })
    }

    fun initPresenterCallView() {
        videoContainer = RelativeLayout(activity)

        val presenterSettings = CometChatCalls.PresentationSettingsBuilder(activity)
            .setDefaultLayoutEnable(call.argument<Boolean>("defaultLayout") ?: true)
            .setIsAudioOnly(call.argument<Boolean>("isAudioOnly") ?: false)
            .setSingleMode(call.argument<Boolean>("isSingleMode") ?: false)
            .showEndCallButton(call.argument<Boolean>("endCallButtonDisable") ?: false)
            .showRecordingButton(call.argument<Boolean>("showRecordingButton") ?: false)
            .showSwitchCameraButton(call.argument<Boolean>("switchCameraButtonDisable") ?: false)
            .showMuteAudioButton(call.argument<Boolean>("muteAudioButtonDisable") ?: false)
            .showPauseVideoButton(call.argument<Boolean>("pauseVideoButtonDisable") ?: false)
            .showAudioModeButton(call.argument<Boolean>("audioModeButtonDisable") ?: false)
            .startWithAudioMuted(call.argument<Boolean>("startAudioMuted") ?: false)
            .startWithVideoMuted(call.argument<Boolean>("startVideoMuted") ?: false)
            .setIsPresenter(call.argument<Boolean>("isPresenter") ?: false)
            .setDefaultAudioMode(call.argument<String>("defaultAudioMode") ?: "SPEAKER")
            .setEventListener(object : CometChatCallsEventsListener {
                override fun onCallEnded() {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.booleanMap(true, "onCallEnded"))
                    }
                }

                override fun onCallEndButtonPressed() {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.booleanMap(true, "onCallEndButtonPressed"))
                    }
                }

                override fun onUserJoined(p0: RTCUser?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcUserMap(p0, "onUserJoined"))
                    }
                }

                override fun onUserLeft(p0: RTCUser?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcUserMap(p0, "onUserLeft"))
                    }
                }

                override fun onUserListChanged(p0: ArrayList<RTCUser>?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcUserListMap(p0, "onUserListChanged"))
                    }
                }

                override fun onAudioModeChanged(p0: ArrayList<AudioMode>?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.audioModeListMap(p0, "onAudioModeChanged"))
                    }
                }

                override fun onCallSwitchedToVideo(p0: CallSwitchRequestInfo?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.callSwitchRequestInfoMap(p0, "onCallSwitchedToVideo"))
                    }
                }

                override fun onUserMuted(p0: RTCMutedUser?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcMutedUserMap(p0, "onUserMuted"))
                    }
                }

                override fun onRecordingToggled(p0: RTCRecordingInfo?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.rtcRecordingInfoMap(p0, "onRecordingToggled"))
                    }
                }

                override fun onError(p0: CometChatException?) {
                    uiThreadHandler.post {
                        callsEventsStreamSink.success(GetMappedObject.cometchatExceptionMap(p0, "onError"))
                    }
                }
            })
            .build()

        val callToken: String ?= call.argument<String>(PluginConstants.CometchatCallsStartSessionArgumentsKeys.CALL_TOKEN)

        CometChatCalls.joinPresentation(callToken, presenterSettings, videoContainer, object : CometChatCalls.CallbackListener<String>(){
            override fun onSuccess(p0: String) {
                uiThreadHandler.post {
                    result.success(GetMappedObject.stringToMap(p0, PluginConstants.Method.SUCCESS))
                }
            }

            override fun onError(e: CometChatException) {
                uiThreadHandler.post {
                    result.error(e.code, e.message, e.details)
                }
            }
        })
    }

    override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
        return CometChatCallsPlatformView(videoContainer)
        //return CometChatCallsPlatformView(parentContainer)
    }
}